### DISCUSSION TITLE
Changelogs for private packages: private repo in GitHub, repository in GCP's Artifact Registry

### DISCUSSION THREAD

**Discussion by jacek-jablonski** [renovatebot/renovate#29211] [Request Help] [answered]
### What would you like help with?

Other

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

Hi,

We are hosting our private packages on GCP Artifact Registry and sources on private GitHub repositories. Packages are prepared using poetry. 

I noticed that changelogs from our private packages are not showing in renovate's PRs. Looking through the logs, I see that for private packages the `sourceUrl` field is empty, which will cause this problem. Where does the value of this `sourceUrl` field come from and what should I complete to make the changelogs populate?

I am aware that I can add the `sourceUrl` field in the renovate config file using the `packageRules` declaration, but I am wondering if I can do it differently by completing this field in the package itself.

### Logs (if relevant)

<details><summary>Logs</summary>

```

{
            "currentValue": "^1.5.0",
            "currentVersion": "1.8.1",
            "datasource": "pypi",
            "depName": "rtb-apps-config",
            "depType": "dependencies",
            "fixedVersion": "1.8.1",
            "lockedVersion": "1.8.1",
            "managerData": {
              "nestedVersion": true
            },
            "packageName": "rtb-apps-config",
            "registryUrl": "https://xxx",
            "versioning": "poetry",
            "warnings": [],
            "updates": []
          },
```

</details>

**Comment by secustor** [comment, marked as answer]
This information is not yet surfaced to the docs but will be soonish. https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/pypi/index.ts#L37

**Reply by jacek-jablonski** [reply]
Thank you.
Unfortunately, it seems that artifact registry doesn't expose json endpoints, so renovate is unable to get `sourceUrl` from repository.

It looks like we'll need to fill in `sourceUrl` in renovate's configuration. 
Assuming that each of our packages has a similar naming scheme:
`ourorg-config`
`ourorg-logging`
`ourorg-metrics`
and has corresponding repositories on GitHub:
https://github.com/myorg/ourorg-config
https://github.com/myorg/ourorg-logging
https://github.com/myorg/ourorg-metrics
can we somehow use regular expressions to do this with one rule instead of listing all the packages in the configuration?

**Reply by jacek-jablonski** [reply]
I tried to do something like:

```
  "packageRules": [
    {
      "matchDepPrefixes": ["ourorg-"],
      "sourceUrl": "https://github.com/myorg/{{depName}}"
    }
]
```

but with no success.

**Reply by secustor** [reply]
We currently not allow setting `sourceUrl` to be set manually.

https://github.com/renovatebot/renovate/discussions/27959

**Reply by jacek-jablonski** [reply]
Hmmm, but it works:
```
  "packageRules": [
    {
      "matchDepNames": ["somepackage"],
      "matchDatasources": ["pypi"],
      "sourceUrl": "https://github.com/myorg/somepackage"
    },
```
and release notes are visible in PR.

**Reply by rarkins** [reply]
It's always been possible to add it unofficially through global packageRules. We'll add official support soon as it does not seem to interfere with our caching

**Reply by jacek-jablonski** [reply]
Would it be possible to make this field `sourceUrl` behave like template?

As Artifact Registry does not provide json endpoints, users of such a registry will always have to list all their packages, which is not very convenient :).

**Reply by rarkins** [reply]
Can you give an example of what your config would look like if you had templating available?

**Reply by jacek-jablonski** [reply]
f.e.:
```
"packageRules": [
  {
    "matchDepPrefixes": ["ourorg-"],
    "sourceUrl": "https://github.com/myorg/{{depName}}"
  }
]
```

**Reply by rarkins** [reply]
https://github.com/renovatebot/renovate/issues/29385
